<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="Using a custom clustering class" href="customize_clustering.html" /><link rel="prev" title="examples" href="examples.html" />

    <meta name="generator" content="sphinx-5.3.0, furo 2022.06.21"/>
        <title>Basic pipeline for exploratory analysis - robustica 0.1.3 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=40978830699223671f4072448e654b5958f38b89" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">robustica 0.1.3 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">robustica 0.1.3 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder=Search name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Modules</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="RobustICA.html"><code class="docutils literal notranslate"><span class="pre">RobustICA</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="InferComponents.html"><code class="docutils literal notranslate"><span class="pre">InferComponents</span></code></a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html"><code class="docutils literal notranslate"><span class="pre">examples</span></code></a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">Basic pipeline for exploratory analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="customize_clustering.html">Using a custom clustering class</a></li>
<li class="toctree-l1"><a class="reference internal" href="infer_components.html">Inferring the number of components</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <div class="section" id="Basic-pipeline-for-exploratory-analysis">
<h1>Basic pipeline for exploratory analysis<a class="headerlink" href="#Basic-pipeline-for-exploratory-analysis" title="Permalink to this heading">#</a></h1>
<div class="section" id="Load-modules">
<h2>Load modules<a class="headerlink" href="#Load-modules" title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">robustica</span> <span class="kn">import</span> <span class="n">RobustICA</span><span class="p">,</span> <span class="n">corrmats</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">sklearn.utils</span> <span class="kn">import</span> <span class="n">parallel_backend</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Load-data-from-Sastry-(2019)">
<h2>Load data from <a class="reference external" href="https://doi.org/10.1038/s41467-019-13483-w">Sastry (2019)</a><a class="headerlink" href="#Load-data-from-Sastry-(2019)" title="Permalink to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">url</span> <span class="o">=</span> <span class="s2">&quot;https://static-content.springer.com/esm/art%3A10.1038</span><span class="si">%2F</span><span class="s2">s41467-019-13483-w/MediaObjects/41467_2019_13483_MOESM4_ESM.xlsx&quot;</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">ExcelFile</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
<span class="n">data</span><span class="o">.</span><span class="n">sheet_names</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;README&#39;,
 &#39;Metadata&#39;,
 &#39;Expression Data&#39;,
 &#39;S matrix&#39;,
 &#39;A matrix&#39;,
 &#39;Media Recipes&#39;,
 &#39;Gene Information&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;README&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Sheet Name</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Experimental Conditions</td>
      <td>Experimental conditions for each sample in PRE...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Expression Data</td>
      <td>Expression levels of genes (log2 Transcripts p...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>S matrix</td>
      <td>I-modulon gene coefficients. Each column is an...</td>
    </tr>
    <tr>
      <th>3</th>
      <td>A matrix</td>
      <td>Condition-specific i-modulon activities. Each ...</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Media Description</td>
      <td>Recipes for base media and trace element mixtures</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Gene Information</td>
      <td>Mapping of b-numbers to gene names/annotations...</td>
    </tr>
    <tr>
      <th>6</th>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Metadata Columns</td>
      <td>Description</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Sample ID</td>
      <td>Unique sample identifier for experiment (e.g. ...</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Study</td>
      <td>Short 1-2 word description of study where samp...</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Project ID</td>
      <td>Identifier for study</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Condition ID</td>
      <td>Identifier for unique experimental conditions</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Replicate #</td>
      <td>Suffix for sample ID to distinguish between re...</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Strain Description</td>
      <td>Describes gene knock-out, mutation knock-in, o...</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Strain</td>
      <td>MG1655 or BW25113</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Base Media</td>
      <td>Media recipe as described in Media Description...</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Carbon Source (g/L)</td>
      <td>Carbon source in g/L</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Nitrogen Source (g/L)</td>
      <td>Nitrogen source in g/L</td>
    </tr>
    <tr>
      <th>18</th>
      <td>Electron Acceptor</td>
      <td>Aerobic (O2), anaerobic (None), or nitrate (KN...</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Trace Element Mixture</td>
      <td>Trace element recipe as described in Media Des...</td>
    </tr>
    <tr>
      <th>20</th>
      <td>Supplement</td>
      <td>Additional media supplements in each experimen...</td>
    </tr>
    <tr>
      <th>21</th>
      <td>Temperature (C)</td>
      <td>Temperature of growth condition in Celsius</td>
    </tr>
    <tr>
      <th>22</th>
      <td>pH</td>
      <td>Neutral (7) or acidic (5.5)</td>
    </tr>
    <tr>
      <th>23</th>
      <td>Antibiotic</td>
      <td>Treatment with low concentration Kanamycin for...</td>
    </tr>
    <tr>
      <th>24</th>
      <td>Culture Type</td>
      <td>Batch or Chemostat (always exponential growth ...</td>
    </tr>
    <tr>
      <th>25</th>
      <td>Growth Rate (1/hr)</td>
      <td>Measured growth rate of strain in media</td>
    </tr>
    <tr>
      <th>26</th>
      <td>Evolved Sample</td>
      <td>Whether the strain was from an adaptive labora...</td>
    </tr>
    <tr>
      <th>27</th>
      <td>Isolate Type</td>
      <td>Clonal or Population (only applicable to ALE s...</td>
    </tr>
    <tr>
      <th>28</th>
      <td>Sequencing Machine</td>
      <td>Type of machine used for sequencing</td>
    </tr>
    <tr>
      <th>29</th>
      <td>Additional Details</td>
      <td>Additional information about experiment</td>
    </tr>
    <tr>
      <th>30</th>
      <td>Biological Replicates</td>
      <td>Number of biological replicates within study</td>
    </tr>
    <tr>
      <th>31</th>
      <td>Alignment</td>
      <td>Percent of reads aligned by bowtie to referenc...</td>
    </tr>
    <tr>
      <th>32</th>
      <td>DOI</td>
      <td>Accession number for study</td>
    </tr>
    <tr>
      <th>33</th>
      <td>GEO</td>
      <td>GEO Accession ID</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Preprocess-data">
<h2>Preprocess data<a class="headerlink" href="#Preprocess-data" title="Permalink to this heading">#</a></h2>
<p>Following the article’s methods, we compute the log2FCs with respect to controls</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># get gene expression</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;Expression Data&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s2">&quot;log-TPM&quot;</span><span class="p">)</span>

<span class="c1"># get controls</span>
<span class="n">controls</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">X</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="s2">&quot;control&quot;</span> <span class="ow">in</span> <span class="n">c</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">controls</span><span class="p">)</span>

<span class="c1"># compute log2FCs with respect to controls</span>
<span class="n">X</span> <span class="o">=</span> <span class="n">X</span> <span class="o">-</span> <span class="n">X</span><span class="p">[</span><span class="n">controls</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="n">X</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;control__wt_glc__1&#39;, &#39;control__wt_glc__2&#39;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>control__wt_glc__1</th>
      <th>control__wt_glc__2</th>
      <th>fur__wt_dpd__1</th>
      <th>fur__wt_dpd__2</th>
      <th>fur__wt_fe__1</th>
      <th>fur__wt_fe__2</th>
      <th>fur__delfur_dpd__1</th>
      <th>fur__delfur_dpd__2</th>
      <th>fur__delfur_fe2__1</th>
      <th>fur__delfur_fe2__2</th>
      <th>...</th>
      <th>efeU__menFentC_ale29__1</th>
      <th>efeU__menFentC_ale29__2</th>
      <th>efeU__menFentC_ale30__1</th>
      <th>efeU__menFentC_ale30__2</th>
      <th>efeU__menFentCubiC_ale36__1</th>
      <th>efeU__menFentCubiC_ale36__2</th>
      <th>efeU__menFentCubiC_ale37__1</th>
      <th>efeU__menFentCubiC_ale37__2</th>
      <th>efeU__menFentCubiC_ale38__1</th>
      <th>efeU__menFentCubiC_ale38__2</th>
    </tr>
    <tr>
      <th>log-TPM</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>b0002</th>
      <td>-0.061772</td>
      <td>0.061772</td>
      <td>0.636527</td>
      <td>0.819793</td>
      <td>-0.003615</td>
      <td>-0.289353</td>
      <td>-1.092023</td>
      <td>-0.777289</td>
      <td>0.161343</td>
      <td>0.145641</td>
      <td>...</td>
      <td>-0.797097</td>
      <td>-0.791859</td>
      <td>0.080114</td>
      <td>0.102154</td>
      <td>0.608180</td>
      <td>0.657673</td>
      <td>0.813105</td>
      <td>0.854813</td>
      <td>0.427986</td>
      <td>0.484338</td>
    </tr>
    <tr>
      <th>b0003</th>
      <td>-0.053742</td>
      <td>0.053742</td>
      <td>0.954439</td>
      <td>1.334385</td>
      <td>0.307588</td>
      <td>0.128414</td>
      <td>-0.872563</td>
      <td>-0.277893</td>
      <td>0.428542</td>
      <td>0.391761</td>
      <td>...</td>
      <td>-0.309105</td>
      <td>-0.352535</td>
      <td>-0.155074</td>
      <td>-0.077145</td>
      <td>0.447030</td>
      <td>0.439881</td>
      <td>0.554528</td>
      <td>0.569030</td>
      <td>0.154905</td>
      <td>0.294799</td>
    </tr>
    <tr>
      <th>b0004</th>
      <td>-0.065095</td>
      <td>0.065095</td>
      <td>-0.202697</td>
      <td>0.119195</td>
      <td>-0.264995</td>
      <td>-0.546017</td>
      <td>-1.918349</td>
      <td>-1.577736</td>
      <td>-0.474815</td>
      <td>-0.495312</td>
      <td>...</td>
      <td>-0.184898</td>
      <td>-0.225615</td>
      <td>0.019575</td>
      <td>0.063986</td>
      <td>0.483343</td>
      <td>0.452754</td>
      <td>0.524828</td>
      <td>0.581878</td>
      <td>0.293239</td>
      <td>0.341040</td>
    </tr>
    <tr>
      <th>b0005</th>
      <td>0.028802</td>
      <td>-0.028802</td>
      <td>-0.865171</td>
      <td>-0.951179</td>
      <td>0.428769</td>
      <td>0.123564</td>
      <td>-1.660351</td>
      <td>-1.531147</td>
      <td>0.240353</td>
      <td>-0.151132</td>
      <td>...</td>
      <td>-0.308221</td>
      <td>-0.581714</td>
      <td>0.018820</td>
      <td>0.004040</td>
      <td>-1.228763</td>
      <td>-1.451750</td>
      <td>-0.839203</td>
      <td>-0.529349</td>
      <td>-0.413336</td>
      <td>-0.478682</td>
    </tr>
    <tr>
      <th>b0006</th>
      <td>0.009087</td>
      <td>-0.009087</td>
      <td>-0.131039</td>
      <td>-0.124079</td>
      <td>-0.144870</td>
      <td>-0.090152</td>
      <td>-0.219917</td>
      <td>-0.046648</td>
      <td>-0.044537</td>
      <td>-0.089204</td>
      <td>...</td>
      <td>1.464603</td>
      <td>1.415706</td>
      <td>1.230831</td>
      <td>1.165153</td>
      <td>0.447447</td>
      <td>0.458852</td>
      <td>0.421417</td>
      <td>0.408077</td>
      <td>1.151066</td>
      <td>1.198529</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>b4688</th>
      <td>-0.261325</td>
      <td>0.261325</td>
      <td>-1.425581</td>
      <td>-2.734490</td>
      <td>0.181893</td>
      <td>0.514395</td>
      <td>-1.943947</td>
      <td>-1.992701</td>
      <td>0.066037</td>
      <td>-0.695325</td>
      <td>...</td>
      <td>-0.885297</td>
      <td>-0.462485</td>
      <td>-2.734490</td>
      <td>-1.451148</td>
      <td>-1.379069</td>
      <td>-1.567420</td>
      <td>-0.999610</td>
      <td>-1.726577</td>
      <td>-2.734490</td>
      <td>-1.189069</td>
    </tr>
    <tr>
      <th>b4693</th>
      <td>-0.278909</td>
      <td>0.278909</td>
      <td>1.361362</td>
      <td>1.020310</td>
      <td>0.608108</td>
      <td>0.988541</td>
      <td>2.558416</td>
      <td>2.142724</td>
      <td>3.120867</td>
      <td>3.104887</td>
      <td>...</td>
      <td>-0.374963</td>
      <td>0.856574</td>
      <td>-1.147824</td>
      <td>-0.814089</td>
      <td>2.054471</td>
      <td>1.853620</td>
      <td>1.957717</td>
      <td>1.943582</td>
      <td>2.233115</td>
      <td>2.023755</td>
    </tr>
    <tr>
      <th>b4696_1</th>
      <td>0.050526</td>
      <td>-0.050526</td>
      <td>1.166436</td>
      <td>1.043373</td>
      <td>-0.531441</td>
      <td>-0.581626</td>
      <td>0.914055</td>
      <td>0.731165</td>
      <td>-0.127269</td>
      <td>-0.164046</td>
      <td>...</td>
      <td>0.261604</td>
      <td>0.278426</td>
      <td>0.201089</td>
      <td>-0.017780</td>
      <td>0.138178</td>
      <td>0.122287</td>
      <td>0.504402</td>
      <td>0.425213</td>
      <td>0.629383</td>
      <td>0.805945</td>
    </tr>
    <tr>
      <th>b4696_2</th>
      <td>-0.031653</td>
      <td>0.031653</td>
      <td>0.785573</td>
      <td>0.881353</td>
      <td>-0.477271</td>
      <td>-0.916095</td>
      <td>0.837603</td>
      <td>0.801393</td>
      <td>-0.071710</td>
      <td>-0.000540</td>
      <td>...</td>
      <td>-0.499371</td>
      <td>0.398783</td>
      <td>0.096609</td>
      <td>-0.103446</td>
      <td>-0.519098</td>
      <td>0.615363</td>
      <td>0.343959</td>
      <td>0.580288</td>
      <td>0.366905</td>
      <td>0.702608</td>
    </tr>
    <tr>
      <th>b4705</th>
      <td>0.724324</td>
      <td>-0.724324</td>
      <td>-4.350151</td>
      <td>-4.317498</td>
      <td>-0.747489</td>
      <td>-1.257045</td>
      <td>-3.308337</td>
      <td>-4.421970</td>
      <td>-2.679693</td>
      <td>-1.872713</td>
      <td>...</td>
      <td>-1.968530</td>
      <td>-1.365300</td>
      <td>-5.468290</td>
      <td>-2.997169</td>
      <td>-3.673367</td>
      <td>-3.161608</td>
      <td>-3.959910</td>
      <td>-4.088644</td>
      <td>-5.468290</td>
      <td>-5.468290</td>
    </tr>
  </tbody>
</table>
<p>3923 rows × 278 columns</p>
</div></div>
</div>
</div>
<div class="section" id="Run-robustica">
<h2>Run <code class="docutils literal notranslate"><span class="pre">robustica</span></code><a class="headerlink" href="#Run-robustica" title="Permalink to this heading">#</a></h2>
<div class="section" id="Using-our-version-of-the-Icasso-algorithm">
<h3>Using our version of the <em>Icasso</em> algorithm<a class="headerlink" href="#Using-our-version-of-the-Icasso-algorithm" title="Permalink to this heading">#</a></h3>
<p>By default, <code class="docutils literal notranslate"><span class="pre">robustica</span></code> performs robust ICA by 1. running <a class="reference external" href="https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.FastICA.html">FastICA</a> multiple times (<code class="docutils literal notranslate"><span class="pre">robust_runs</span></code>=100) 2. inferring and correcting the signs of components across runs (<code class="docutils literal notranslate"><span class="pre">robust_infer_signs=True</span></code>) 3. compressing the feature space (rows) (<code class="docutils literal notranslate"><span class="pre">robust_dimreduce=True</span></code>) 4. using sign-corrected and compressed components to obtain clusters of independent components (<code class="docutils literal notranslate"><span class="pre">robust_method=&quot;AgglomerativeClustering&quot;</span></code>
and <code class="docutils literal notranslate"><span class="pre">robust_kws={&quot;linkage&quot;:</span> <span class="pre">&quot;average&quot;}</span></code>) 5. uses the clustering labels to compute robust independent components</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>

<span class="n">rica</span> <span class="o">=</span> <span class="n">RobustICA</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">robust_runs</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="k">with</span> <span class="n">parallel_backend</span><span class="p">(</span><span class="s2">&quot;loky&quot;</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">):</span>
    <span class="n">S</span><span class="p">,</span> <span class="n">A</span> <span class="o">=</span> <span class="n">rica</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">S</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">A</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/8 [00:00&lt;?, ?it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running FastICA multiple times...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 8/8 [00:03&lt;00:00,  2.24it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Inferring sign of components...
Reducing dimensions...
Clustering...
Computing centroids...
(3923, 100) (278, 100)
CPU times: user 3.71 s, sys: 312 ms, total: 4.02 s
Wall time: 7.6 s
</pre></div></div>
</div>
</div>
<div class="section" id="Using-the-classical-Icasso-algorithm">
<h3>Using the classical <em>Icasso</em> algorithm<a class="headerlink" href="#Using-the-classical-Icasso-algorithm" title="Permalink to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%time</span>

<span class="n">rica</span> <span class="o">=</span> <span class="n">RobustICA</span><span class="p">(</span>
    <span class="n">n_components</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">robust_runs</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span>
    <span class="n">robust_infer_signs</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">robust_dimreduce</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">robust_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;affinity&quot;</span><span class="p">:</span> <span class="s2">&quot;precomputed&quot;</span><span class="p">,</span> <span class="s2">&quot;linkage&quot;</span><span class="p">:</span> <span class="s2">&quot;average&quot;</span><span class="p">},</span>
<span class="p">)</span>

<span class="k">with</span> <span class="n">parallel_backend</span><span class="p">(</span><span class="s2">&quot;loky&quot;</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">):</span>
    <span class="n">S_classic</span><span class="p">,</span> <span class="n">A_classic</span> <span class="o">=</span> <span class="n">rica</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">S_classic</span><span class="o">.</span><span class="n">shape</span><span class="p">,</span> <span class="n">A_classic</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/8 [00:00&lt;?, ?it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running FastICA multiple times...
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 8/8 [00:02&lt;00:00,  3.09it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Precomputing distance matrix...
Clustering...
Computing centroids...
(3923, 100) (278, 100)
CPU times: user 2.38 s, sys: 132 ms, total: 2.51 s
Wall time: 7.51 s
</pre></div></div>
</div>
</div>
<div class="section" id="Correspondence-between-the-two-approaches">
<h3>Correspondence between the two approaches<a class="headerlink" href="#Correspondence-between-the-two-approaches" title="Permalink to this heading">#</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">11</span><span class="p">,</span> <span class="mi">10</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">corrmats</span><span class="p">(</span><span class="n">S</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">S_classic</span><span class="o">.</span><span class="n">T</span><span class="p">)),</span>
    <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span>
    <span class="n">center</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">cbar_kws</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;label&quot;</span><span class="p">:</span> <span class="sa">r</span><span class="s2">&quot;$abs(Pearson Correlation)$&quot;</span><span class="p">},</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Classic ICs&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Default ICs&quot;</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/basics_14_0.png" src="_images/basics_14_0.png" />
</div>
</div>
</div>
</div>
</div>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="customize_clustering.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">Using a custom clustering class</div>
              </div>
              <svg><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="examples.html">
              <svg><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title"><code class="docutils literal notranslate"><span class="pre">examples</span></code></div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2022, Miquel Anglada Girotto
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            <div class="icons">
              
            </div>
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            Contents
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">Basic pipeline for exploratory analysis</a><ul>
<li><a class="reference internal" href="#Load-modules">Load modules</a></li>
<li><a class="reference internal" href="#Load-data-from-Sastry-(2019)">Load data from Sastry (2019)</a></li>
<li><a class="reference internal" href="#Preprocess-data">Preprocess data</a></li>
<li><a class="reference internal" href="#Run-robustica">Run <code class="docutils literal notranslate"><span class="pre">robustica</span></code></a><ul>
<li><a class="reference internal" href="#Using-our-version-of-the-Icasso-algorithm">Using our version of the <em>Icasso</em> algorithm</a></li>
<li><a class="reference internal" href="#Using-the-classical-Icasso-algorithm">Using the classical <em>Icasso</em> algorithm</a></li>
<li><a class="reference internal" href="#Correspondence-between-the-two-approaches">Correspondence between the two approaches</a></li>
</ul>
</li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["\\(", "\\)"]], "displayMath": [["\\[", "\\]"]], "processRefs": false, "processEnvironments": false}})</script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    </body>
</html>